base/null.jl

```diff
@@ -0,0 +1,116 @@
+# This file is a part of Julia. License is MIT: https://julialang.org/license
+
+"""
+    Null
+
+A type with no fields that is the type [`null`](@ref).
+"""
+struct Null end
+
+"""
+    null
+
+The singleton instance of type [`Null`](@ref), used to denote a missing value.
+"""
+const null = Null()
+
+"""
+    Some{T}
+
+A wrapper type used with [`Option`](@ref) (a.k.a. `Union{Null, Some{T}}`). Its main
+purposes are to force handling the possibility of the `Option` being [`null`](@ref)
+by calling [`unwrap`](@ref) (with [`isnull`](@ref)) before actually using the wrapped
+value; and to allow distinguishing `null` from `Some(null)` when an `Option` may wrap
+a `null` value.
+"""
+struct Some{T}
+    value::T
+end
+
+"""
+    Option{T}
+
+A type alias for `Union{Null, Some{T}}` used to hold either a value of type `T` inside
+a [`Some`](@ref) wrapper, or a [`null`](@ref) value. Use [`isnull`](@ref) to check whether
+an `Option` object is null, and [`unwrap`](@ref) to access the value inside the `Some`
+object if not.
+"""
+Option{T} = Union{Null, Some{T}}
+
+eltype(::Type{Some{T}}) where {T} = T
+
+promote_rule(::Type{Some{S}}, ::Type{Some{T}}) where {S,T} = Some{promote_type(S, T)}
+promote_rule(::Type{Some{T}}, ::Type{Null}) where {T} = Option{T}
+
+convert(::Type{Some{T}}, x::Some) where {T} = Some{T}(convert(T, x.value))
+
+convert(::Type{Null}, ::Null) = null
+convert(::Type{Null}, ::Void) = null
+convert(::Type{Null}, ::Any) = throw(NullException())
+
+convert(::Type{Option{T}}, ::Null) where {T} = null
+convert(::Type{Option   }, ::Null)           = null
+# FIXME: find out why removing these two methods makes addprocs() fail
```

Comment:
This has turned out to be very hard to debug. Without these two methods, `addprocs(2)` hangs and finally times out, leaving two julia processes using 100% CPU in the background. Replacing all occurrences of `nothing` with `null` in base/distributed/* didn't fix it. The output isn't very helpful. Any help on how to debug this is welcome!

```julia
julia> addprocs(2)
ERROR: Timed out waiting to read host:port string from worker.
read_worker_host_port(::Pipe) at ./distributed/cluster.jl:302
connect(::Base.Distributed.LocalManager, ::Int64, ::WorkerConfig) at ./distributed/managers.jl:395
create_worker(::Base.Distributed.LocalManager, ::WorkerConfig) at ./distributed/cluster.jl:507
setup_launched_worker(::Base.Distributed.LocalManager, ::WorkerConfig, ::Array{Int64,1}) at ./distributed/cluster.jl:453
(::getfield(Base.Distributed, Symbol("##41#44")){Base.Distributed.LocalManager,WorkerConfig})() at ./task.jl:335

...and 1 more exception(s).

Stacktrace:
 [1] sync_end() at ./task.jl:287
 [2] macro expansion at ./task.jl:303 [inlined]
 [3] #addprocs_locked#38(::Array{Any,1}, ::Function, ::Base.Distributed.LocalManager) at ./distributed/cluster.jl:407
 [4] #addprocs#37(::Array{Any,1}, ::Function, ::Base.Distributed.LocalManager) at ./distributed/cluster.jl:371
 [5] #addprocs#257(::Bool, ::Array{Any,1}, ::Function, ::Int64) at ./distributed/managers.jl:314
 [6] addprocs(::Int64) at ./distributed/managers.jl:313
```

Reply:
Perhaps related to automatic conversion thru fallback constructor? https://github.com/JuliaLang/julia/pull/23273

Reply:
Generally this rather happens when assigning to an `Option` field. It's quite rare to call `Option(x)`.

I've finally found a way to see the backtrace from the workers, by adding print statements to `read_worker_host_port`. Turns out the `TCPSocket` inner constructor used `nothing` to set a `Nullable` field to null. Now it works!